File: 17. Letter Combinations of a Phone Number/17. Letter Combinations of a Phone Number.py
from typing import List


class Solution:
    def letterCombinations(self, digits: str) -> List[str]:
        if digits == "":
            return []

        digits_map = {
            "2": "abc",
            "3": "def",
            "4": "ghi",
            "5": "jkl",
            "6": "mno",
            "7": "pqrs",
            "8": "tuv",
            "9": "wxyz"
        }

        n = len(digits)
        ans, curr = [], []

        def backtrack(i):
            if len(curr) == n:
                ans.append("".join(curr))
                return

            for c in digits_map[digits[i]]:
                curr.append(c)
                backtrack(i + 1)
                curr.pop()

        backtrack(0)
        return ans

File: 17. Letter Combinations of a Phone Number/17. Letter Combinations of a Phone Number.js
/**
 * @param {string} digits
 * @return {string[]}
 */
var letterCombinations = function (digits) {
  if (digits === "") {
    return [];
  }

  const digitsMap = {
    2: "abc",
    3: "def",
    4: "ghi",
    5: "jkl",
    6: "mno",
    7: "pqrs",
    8: "tuv",
    9: "wxyz",
  };

  const n = digits.length;
  const ans = [];
  const curr = [];

  const backtrack = (i) => {
    if (curr.length === n) {
      ans.push(curr.join(""));
      return;
    }

    for (const c of digitsMap[digits[i]]) {
      curr.push(c);
      backtrack(i + 1);
      curr.pop();
    }
  };

  backtrack(0);
  return ans;
};
